File: ios/BridgeModule.m
//
//  BridgeModule.m
//  ResonanceAudioDummyApp
//
//  Created by Techtic on 24/06/22.
//

#import "BridgeModule.h"
#import <React/RCTLog.h>
#import "GVRAudioEngine.h"

@implementation BridgeModule
RCT_EXPORT_MODULE(BridgeModule);

// Sample sound file names.
GVRAudioEngine *_gvr_audio_engine;
static NSString kObjectSoundFile = "cube_sound.wav";
int _sound_object_id = _gvr_audio_engine.INVALID_ID;

RCT_EXPORT_METHOD(playAudio)
{
  [_gvr_audio_engine playSound:_sound_object_id loopingEnabled:true];
  [_gvr_audio_engine preloadSoundFile: kObjectSoundFile];
  //Start audio playback
  [_gvr_audio_engine start];
}

RCT_EXPORT_METHOD(stopAudio:(NSString *)filePath)
{
  kObjectSoundFile filePath;
  [_gvr_audio_engine stopSound: _sound_object_id];
  _sound_object_id = [_gvr_audio_engine createSoundObject:kObjectSoundFile];
}

RCT_EXPORT_METHOD(changePosition:(float *)x y:(float *)y z:(float *)z )
{
 [_gvr_audio_engine setSoundObjectPosition:_sound_object_id
                                          x:x
                                          y:y
                                          z:z];
}

RCT_EXPORT_METHOD(setFilePath:(NSString *)filePath )
{
  kObjectSoundFile = filePath;
  // Initialize the GVRAudioEngine with high binaural rendering quality.
  // High binaural quality is the default rendering mode.
  _gvr_audio_engine =
      [[GVRAudioEngine alloc] initWithRenderingMode:kRenderingModeBinauralHighQuality];
  // Preload sound files.
  // This method can be called on mono and multi-channel Ambisonic sound files.
  [_gvr_audio_engine preloadSoundFile:kObjectSoundFile];
  // Create the first sound object
  _sound_object_id = [_gvr_audio_engine createSoundObject:kObjectSoundFile];
}
@end

File: android/app/src/main/java/com/resonanceaudiodummyapp/BridgeModule.java
package com.resonanceaudiodummyapp;


import androidx.annotation.NonNull;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.google.vr.sdk.audio.GvrAudioEngine;

public class BridgeModule extends ReactContextBaseJavaModule {
    BridgeModule(ReactApplicationContext context) {
        super(context);
    }

    private GvrAudioEngine gvrAudioEngine;
    private static String OBJECT_SOUND_FILE = "cube_sound.wav";
    private volatile int sourceId = GvrAudioEngine.INVALID_ID;

    @NonNull
    @Override
    public String getName() {
        return "BridgeModule";
    }

    @ReactMethod
    public void playAudio() {
        new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        gvrAudioEngine.playSound(sourceId, true /* looped playback */);
                        // Preload an unspatialized sound to be played on a successful trigger on the cube.
                        gvrAudioEngine.preloadSoundFile(OBJECT_SOUND_FILE);
                    }
                })
                .start();
    }

    @ReactMethod
    public void stopAudio() {
        gvrAudioEngine.stopSound(sourceId);
           gvrAudioEngine = new GvrAudioEngine(getReactApplicationContext(), GvrAudioEngine.RenderingMode.BINAURAL_HIGH_QUALITY);
        gvrAudioEngine.preloadSoundFile(OBJECT_SOUND_FILE);
        sourceId = gvrAudioEngine.createSoundObject(OBJECT_SOUND_FILE);
    }

    @ReactMethod
    public void changePosition(float x, float y, float z) {
        gvrAudioEngine.setSoundObjectPosition(sourceId, x, y, z);
    }

    @ReactMethod
    public void setFilePath(String filePath)  {
        OBJECT_SOUND_FILE = filePath;
        gvrAudioEngine = new GvrAudioEngine(getReactApplicationContext(), GvrAudioEngine.RenderingMode.BINAURAL_HIGH_QUALITY);
        gvrAudioEngine.preloadSoundFile(OBJECT_SOUND_FILE);
        sourceId = gvrAudioEngine.createSoundObject(OBJECT_SOUND_FILE);
    }
}